// elemental/lib/components/Pagination.js
'use strict';

var React = require('react');
var classNames = require('classnames');

module.exports = React.createClass({
	displayName: 'Pagination',
	propTypes: {
		className: React.PropTypes.string,
		currentPage: React.PropTypes.number.isRequired,
		onPageSelect: React.PropTypes.func,
		pageSize: React.PropTypes.number.isRequired,
		plural: React.PropTypes.string,
		singular: React.PropTypes.string,
		style: React.PropTypes.object,
		total: React.PropTypes.number.isRequired,
		limit: React.PropTypes.number
	},
	renderCount: function renderCount() {
		var count = '';
		var _props = this.props;
		var currentPage = _props.currentPage;
		var pageSize = _props.pageSize;
		var plural = _props.plural;
		var singular = _props.singular;
		var total = _props.total;

		if (!total) {
			count = 'No ' + (plural || 'records');
		} else if (total > pageSize) {
			var start = pageSize * (currentPage - 1) + 1;
			var end = Math.min(start + pageSize - 1, total);
			count = 'Showing ' + start + ' to ' + end + ' of ' + total;
		} else {
			count = 'Showing ' + total;
			if (total > 1 && plural) {
				count += ' ' + plural;
			} else if (total === 1 && singular) {
				count += ' ' + singular;
			}
		}
		return React.createElement(
			'div',
			{ className: 'Pagination__count' },
			count
		);
	},
	onPageSelect: function onPageSelect(i) {
		if (!this.props.onPageSelect) return;
		this.props.onPageSelect(i);
	},
	renderPages: function renderPages() {
		var _this = this;

		if (this.props.total <= this.props.pageSize) return null;

		var pages = [];
		var _props2 = this.props;
		var currentPage = _props2.currentPage;
		var pageSize = _props2.pageSize;
		var total = _props2.total;
		var limit = _props2.limit;

		var totalPages = Math.ceil(total / pageSize);
		var minPage = 1;
		var maxPage = totalPages;

		if (limit && limit < totalPages) {
			var rightLimit = Math.floor(limit / 2);
			var leftLimit = rightLimit + limit % 2 - 1;

			minPage = currentPage - leftLimit;
			maxPage = currentPage + rightLimit;

			if (minPage < 1) {
				maxPage = limit;
				minPage = 1;
			}

			if (maxPage > totalPages) {
				minPage = totalPages - limit + 1;
				maxPage = totalPages;
			}
		}

		if (minPage > 1) {
			pages.push(React.createElement(
				'button',
				{ key: 'page_start', className: 'Pagination__list__item', onClick: function () {
						return _this.onPageSelect(1);
					} },
				'...'
			));
		}

		var _loop = function (page) {
			var current = page === currentPage;
			var className = classNames('Pagination__list__item', {
				'is-selected': current
			});
			/* eslint-disable no-loop-func */
			pages.push(React.createElement(
				'button',
				{ key: 'page_' + page, className: className, onClick: function () {
						return _this.onPageSelect(page);
					} },
				page
			));
			/* eslint-enable */
		};

		for (var page = minPage; page <= maxPage; page++) {
			_loop(page);
		}

		if (maxPage < totalPages) {
			pages.push(React.createElement(
				'button',
				{ key: 'page_end', className: 'Pagination__list__item', onClick: function () {
						return _this.onPageSelect(totalPages);
					} },
				'...'
			));
		}

		return React.createElement(
			'div',
			{ className: 'Pagination__list' },
			pages
		);
	},
	render: function render() {
		var className = classNames('Pagination', this.props.className);
		return React.createElement(
			'div',
			{ className: className, style: this.props.style },
			this.renderCount(),
			this.renderPages()
		);
	}
});
